### PR TITLE
create gesture recognizers on attach and dispose on detach to avoid leaks

### LINKED ISSUE #39636: RenderEditable leaks its gesture recognizers
RenderEditable creates gesture recognizers on creation, and never disposes them. It should create them on attach, and dispose them on detach.

### PR DESCRIPTION
**Pull request by jaumard**
## Description

Create gesture recognizer on attach and dispose them on dispose to avoid leaks on RenderEditable

## Related Issues

fix https://github.com/flutter/flutter/issues/39636

## Tests

I added the following tests:

Didn't added any as it's only internal fields, let me know if I can add some and how.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by chunhtai** [comment]
First of all, Welcome and thank you for contributing to flutter !

The change looks good, but we would like to have test to make sure no one break it in the future. Can you add a test for this change?

**Comment by jaumard** [comment]
Thanks :)
Like I said on the description, I don't mind adding test, but I don't know how/what to test here. It's only internal stuff on the class that has moved and wasn't tested previously.
Can you explain me what kind of tests are expected ? I would gladly add them

**Comment by chunhtai** [comment]
Yes, it seems extreme hard to test this method.... Someone might have a better idea

I can only think of an indirect way of testing this
```dart
test('editale test', () {
    final TextSelectionDelegate delegate = FakeEditableTextState();
    final RenderEditable editable = RenderEditable(
      backgroundCursorColor: Colors.grey,
      selectionColor: Colors.black,
      textDirection: TextDirection.ltr,
      cursorColor: Colors.red,
      offset: ViewportOffset.zero(),
      textSelectionDelegate: delegate,
      text: const TextSpan(
        text: 'test',
        style: TextStyle(
          height: 1.0, fontSize: 10.0, fontFamily: 'Ahem',
        ),
      ),
      startHandleLayerLink: LayerLink(),
      endHandleLayerLink: LayerLink(),
      selection: const TextSelection(
        baseOffset: 0,
        extentOffset: 3,
        affinity: TextAffinity.upstream,
      ),
      onSelectionChanged: (_, __, ___) { },
    );
    final PipelineOwner owner = PipelineOwner(onNeedVisualUpdate: () { });
    // Implement CustomTestWidgetsFlutterBinding that implement a spy into
    // GestureBinding.instance.pointerRouter you can use TestWidgetsFlutterBinding
    // as template
    CustomTestWidgetsFlutterBinding.ensureInitialized();
    editable.attach(owner);
    // This should register pointer into GestureBinding.instance.pointerRouter.
    editable.handleEvent(const PointerDownEvent(), BoxHitTestEntry(editable, const Offset(10,10)));
    editable.detach();
    // Make sure GestureBinding.instance.pointerRouter is removed
    // ...
  });
```

When Rendereditable handle a pointer event it will use taprecognizer to register the pointer into GestureBinding.instance.pointerRouter.
In theory we can check GestureBinding.instance.pointerRouter is empty after rendereditable is detached. However, there is no public api to spy into what is in the tracked pointer in GestureBinding.instance.pointerRouter. What you can do is to override the GestureBinding that can spy into GestureBinding.instance.pointerRouter.

**Comment by chunhtai** [comment]
By the way, this test should live in https://github.com/flutter/flutter/blob/master/packages/flutter/test/rendering/editable_test.dart

File: packages/flutter/test/rendering/editable_test.dart [hunk 1]
```diff
@@ -587,4 +592,58 @@ void main() {
...
+    // This should register pointer into GestureBinding.instance.pointerRouter.
+    editable.handleEvent(const PointerDownEvent(), BoxHitTestEntry(editable, const Offset(10,10)));
+    editable.detach();
+    // Make sure GestureBinding.instance.pointerRouter is removed
+
+    // ...
+  });
+}
+
+class GestureBindingSpy extends AutomatedTestWidgetsFlutterBinding {
```
Comment (jaumard) [inline packages/flutter/test/rendering/editable_test.dart]: @chunhtai I'm trying to do that test but I'm quite lost :/ with this class I'm supposed to intercept the event but look like my overrides are never called. 

If I breakpoint before detach I can see the pointer has one route, and then 0 after the call. But it still unclear to me how testing that by code :/ 

Any help tips you can give me ?

**Comment by jaumard** [comment]
@chunhtai ok I finally was able to make that test !! I was struggling to spy the GestureBinding. Let me know if it's ok

File: packages/flutter/test/rendering/editable_test.dart [hunk 2]
```diff
@@ -587,4 +592,64 @@ void main() {
...
+        extentOffset: 3,
+        affinity: TextAffinity.upstream,
+      ),
+      onSelectionChanged: (_, __, ___) { },
+    );
+    final PipelineOwner owner = PipelineOwner(onNeedVisualUpdate: () { });
+    editable.attach(owner);
+    // This should register pointer into GestureBinding.instance.pointerRouter.
+    editable.handleEvent(const PointerDownEvent(), BoxHitTestEntry(editable, const Offset(10,10)));
+    // ignore: avoid_as
```
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: We generally avoid using as cast, you can re cast it first before doing the count
```dart
final _PointerRouterSpy spy = GestureBinding.instance.pointerRouter
expect(spy. routeCount, greaterThan(0));
editable.detach();
expect(spy. routeCount, 0);
```

File: packages/flutter/test/rendering/editable_test.dart [hunk 3]
```diff
@@ -587,4 +592,64 @@ void main() {
...
+    expect((GestureBinding.instance.pointerRouter as _PointerRouterSpy).routeCount, 0);
+  });
+}
+
+class _GestureBindingSpy extends AutomatedTestWidgetsFlutterBinding {
+  final PointerRouter _testPointerRouter = _PointerRouterSpy();
+
+  @override
+  PointerRouter get pointerRouter {
```
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: even better
```dart
PointerRouter get pointerRouter => _testPointerRouter;
```

File: packages/flutter/test/rendering/editable_test.dart [hunk 4]
```diff
@@ -2,11 +2,16 @@
...
+import 'dart:typed_data';
+
+import 'dart:ui' as ui;
+import 'package:flutter/gestures.dart';
 import 'package:flutter/rendering.dart';
 import 'package:flutter_test/flutter_test.dart';
 import 'package:flutter/services.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/foundation.dart';
+import 'package:quiver/time.dart';
```
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: Some of the import are unused?

**Review by chunhtai — COMMENTED** [review]
Nice test!
I left some minor styling comment. overall all looks good.

File: packages/flutter/test/rendering/editable_test.dart (hunk 3 above)
Comment (jaumard) [reply packages/flutter/test/rendering/editable_test.dart]: Ho yes :) I put bracket to be able to put a breakpoint on it ^^ forget to put the arrow style later . Thanks

File: packages/flutter/test/rendering/editable_test.dart (hunk 4 above)
Comment (jaumard) [reply packages/flutter/test/rendering/editable_test.dart]: Ha ha yes I try so many things :) forget to clean up after

**Comment by jaumard** [comment]
@chunhtai I have a problem, look like _GestureBindingSpy() have to be called before the test start, but if I do so all other tests are failing. 
What I need is a `setUp(() => _GestureBindingSpy());` but only for my test and not the other.
Is there a way to achieve this ? Or should I move my tests into it's own file (`editable_gesture_test.dart`) ? Let me know :)

**Comment by jaumard** [comment]
I've push it on another file to be sure all the CI builds were green :) all good

File: packages/flutter/test/rendering/editable_gesture_test.dart [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class FakeEditableTextState extends TextSelectionDelegate {
```
Comment (goderbauer) [inline packages/flutter/test/rendering/editable_gesture_test.dart]: nit: move this to the end of the file (or you might be able to remove it entirely by using a Mockito mock).

File: packages/flutter/test/rendering/editable_test.dart [hunk 6]
```diff
@@ -2,11 +2,12 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:flutter/foundation.dart';
```
Comment (goderbauer) [inline packages/flutter/test/rendering/editable_test.dart]: What are the changes in this file for? Maybe revert?

**Review by goderbauer — COMMENTED** [review]
Looks good with two nits.

File: packages/flutter/test/rendering/editable_test.dart (hunk 6 above)
Comment (jaumard) [reply packages/flutter/test/rendering/editable_test.dart]: Yes in the end the is no changes on that file as I moved my test, I'll revert :)

File: packages/flutter/test/rendering/editable_gesture_test.dart (hunk 5 above)
Comment (jaumard) [reply packages/flutter/test/rendering/editable_gesture_test.dart]: In fact it's the same one as the one under `editable_test.dart` so I can just used that one instead of duplicate it.

Comment (jaumard) [reply packages/flutter/test/rendering/editable_gesture_test.dart]: Ok look like it's not allowed ^^ changing to mockito mock then :)

**Comment by jaumard** [comment]
@goderbauer I have made the changes, but look like some builds need to be restart (errors doesn't look like happening because of my changes)

**Review by goderbauer — APPROVED** [review]
LGTM

I restarted the build for you.

**Comment by Piinks** [comment]
Thank you again for the contribution @jaumard ! 🎉
